File: trunk/quote/Projects/PluginModelIO/QuoteSaver.vb
Imports System.Data.SqlClient
Imports System.Reflection
Imports System.Data.OleDb
Imports System.Transactions
Imports System.Windows.Forms

Imports Model.Quote
Imports Model

Imports DB.QuoteDataBase
Imports DB.QuoteDataBaseTableAdapters

Public Class QuoteSaver

    Public Class QuoteInfoClass
        Public Property PartNumber As String
        Public Property RFQ As String
        Public Property Initials As String
    End Class

    Public Function Save(ByVal q As Model.BOM.Header, _
                         ByVal info As QuoteInfoClass) As Integer
        Return Save(q, info, False)
    End Function

    Public Function Save(ByVal q As Model.BOM.Header, _
                         ByVal info As QuoteInfoClass, _
                         ByVal IsQuote As Boolean) _
                        As Integer

        System.Windows.Forms.Cursor.Current = Cursors.WaitCursor

        ' Ensure the properies are updated
        PluginHost.App.MainForm.Focus()

        Dim o As Model.BOM.PrimaryPropeties = q.PrimaryProperties

        Dim newId As Integer
        Dim id As Integer = o.CommonID
        If IsQuote Then
            id = 0
        End If

        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuoteTableAdapter
        If id > 0 Then
            adaptor.Update( _
                o.Customer.Name, info.RFQ, info.PartNumber, _
                False, q.ID, o.Initials, q.PrimaryProperties.CommonCreatedDate, _
                Date.Now, o.Customer.ID, o.CommonID, o.CommonID)
            newId = id
        Else
            adaptor.Connection.Open()
            Dim proxy As New DB.QuoteTableProxy(adaptor)
            proxy.Transaction = adaptor.Connection.BeginTransaction
            adaptor.Insert(o.Customer.Name, _
                info.RFQ, info.PartNumber, IsQuote, _
                q.ID, info.Initials, Date.Now, Date.Now, _
                o.Customer.ID)
            Dim cmd As SqlCommand = New SqlCommand( _
                "SELECT @@IDENTITY", adaptor.Connection)
            cmd.Transaction = proxy.Transaction
            newId = CInt(cmd.ExecuteScalar())
            proxy.Transaction.Commit()
            adaptor.Connection.Close()
            If id = 0 And Not IsQuote Then
                q.PrimaryProperties.SetID(newId)
            End If
        End If

        adaptor.Connection.Open()
        CommonSaver.DeleteProperties(newId)
        CommonSaver.SaveOtherProperties(newId, q.OtherProperties, True)
        CommonSaver.SaveComputationProperties(newId, _
            q.ComputationProperties, True)
        CommonSaver.SaveNoteProperties(newId, q.NoteProperties)
        CommonSaver.DeleteComponents(newId)
        CommonSaver.SaveComponents(q, newId, True)
        adaptor.Connection.Close()

        System.Windows.Forms.Cursor.Current = Cursors.Default

        Return newId
    End Function

End Class

File: quote/Projects/PluginModelIO/QuoteSaver.cs
namespace Model.IO
{
    using System;
    using System.Data.SqlClient;
    using System.Globalization;
    using System.Linq;
    using System.Windows.Forms;

    public static class QuoteSaver
    {
        public static int Save(
            Model.Template.Header header, 
            QuoteSaverQuoteInfoClass info)
        {
            return Save(header, info, false);
        }

        public static int Save(
            Model.Template.Header header, 
            QuoteSaverQuoteInfoClass info, 
            bool headerIsAQuote)
        {
            System.Windows.Forms.Cursor.Current = Cursors.WaitCursor;

            // Ensure the properies are updated
            Host.App.MainForm.Focus();

            Model.Template.PrimaryProperties o = 
                header.PrimaryProperties as Model.Template.PrimaryProperties;

            int newId = 0;
            int id = o.CommonId;
            
            if (headerIsAQuote) 
            {
                id = 0;
            }

            DB.QuoteDataBaseTableAdapters._QuoteTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuoteTableAdapter();
            
            if (id > 0) 
            {
                adaptor.Update(
                    o.Customer.Name, 
                    info.RequestForQuote, 
                    info.PartNumber, 
                    false,
                    header.Id, 
                    o.Initials, 
                    header.PrimaryProperties.CommonCreatedDate, 
                    System.DateTime.Now, 
                    o.Customer.Id, 
                    o.CommonId,
                    o.CommonId);
                
                newId = id;
            } 
            else 
            {
                adaptor.Connection.Open();
                DB.QuoteTableProxy proxy = new DB.QuoteTableProxy(adaptor);
                proxy.Transaction = adaptor.Connection.BeginTransaction();
                
                adaptor.Insert(
                    o.Customer.Name, 
                    info.RequestForQuote, 
                    info.PartNumber, 
                    headerIsAQuote, 
                    header.Id, 
                    info.Initials, 
                    System.DateTime.Now, 
                    System.DateTime.Now, 
                    o.Customer.Id);
                
                SqlCommand cmd = new SqlCommand("SELECT @@IDENTITY", adaptor.Connection);
                cmd.Transaction = proxy.Transaction;
                newId = Convert.ToInt32(cmd.ExecuteScalar(), CultureInfo.CurrentCulture);
                
                proxy.Transaction.Commit();
                adaptor.Connection.Close();
                
                if (id == 0 & !headerIsAQuote) 
                {
                    header.PrimaryProperties.SetId(newId);
                }
            }

            adaptor.Connection.Open();
            CommonSaver.DeleteProperties(newId);
            CommonSaver.SaveOtherProperties(newId, header.OtherProperties, true);
            CommonSaver.SaveComputationProperties(newId, header.ComputationProperties, true);
            CommonSaver.SaveNoteProperties(newId, header.NoteProperties);
            CommonSaver.DeleteComponents(newId);
            CommonSaver.SaveComponents(header, newId, true);
            adaptor.Connection.Close();

            System.Windows.Forms.Cursor.Current = Cursors.Default;

            return newId;
        }
    }
}
